### PR TITLE
[BugFix] Fix missing node normalization for link prediction tasks in multi-task learning

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
In multitask learning, when there is a training link prediction task with contrastive loss, the loss may become NaN. This is because, GraphStorm does not add proper node normalization for the gnn embeddings.

*Description of changes:*
Fix the bug.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/model/multitask_gnn.py [hunk 1]
```diff
@@ -93,10 +101,102 @@ def __init__(self, alpha_l2norm):
...
+        if self._node_embed_norm_method[task_id] is not None:
+            rank = get_rank()
+            world_size = get_world_size()
+            for key, emb in embs.items():
+                if isinstance(emb, (dgl.distributed.DistTensor, LazyDistTensor)):
+                    # If emb is a distributed tensor, multiple processes are doing
+                    # embdding normalization concurrently. We need to split
+                    # the task. (From full_graph_inference)
+                    start, end = get_data_range(rank, world_size, len(emb))
+                    new_embs = emb if inplace else \
```
Comment (zhjwy9343) [inline python/graphstorm/model/multitask_gnn.py:136]: if inplace=False, may create a new dictionary, e.g., `new_embs = {}` and assign values like 
`new_emb = create_dist_tensor()`
`new_embs[key] = new_emb`

As the doc strings said, the `new_embs` should be a dictionary of tensors. But the line 149 shows it is a tensor.

File: python/graphstorm/model/multitask_gnn.py [hunk 2]
```diff
@@ -93,10 +101,102 @@ def __init__(self, alpha_l2norm):
...
+        dict of Tensors:
+            Normalized node embeddings.
+        """
+        if self._node_embed_norm_method[task_id] is not None:
+            return normalize_node_embs(embs, self._node_embed_norm_method[task_id])
+        else:
+            return embs
+
+    @property
+    def node_embed_norm_methods(self):
```
Comment (zhjwy9343) [inline python/graphstorm/model/multitask_gnn.py:187]: "node_embed_norm_methods" -> "node_embed_norm_method"? Why need an "s" here?

Comment (classicsong) [reply python/graphstorm/model/multitask_gnn.py:187]: As different ask may use different node normalization method.

File: tests/end2end-tests/graphstorm-mt/mgpu_test.sh [hunk 3]
```diff
@@ -674,6 +674,49 @@ python3 $GS_HOME/tests/end2end-tests/check_infer.py --train-embout /data/gsgnn_m
...
+then
+    echo "We use SageMaker task tracker, we should have save_embed_path"
+    exit -1
+fi
+
+cnt=$(ls -l /data/gsgnn_mt/emb/ | wc -l)
+cnt=$[cnt - 1]
+if test $cnt != 3
+then
+    echo "The number of saved embs $cnt is not equal to 3. Two for movie and user and One for link-prediction-subtask normalized embedding."
```
Comment (zhjwy9343) [inline tests/end2end-tests/graphstorm-mt/mgpu_test.sh:697]:
```suggestion
    echo "The number of saved embs $cnt is not equal to 3. Should have Two for movie and user and One for link-prediction-subtask normalized embedding."
```

File: tests/end2end-tests/graphstorm-mt/mgpu_test.sh [hunk 4]
```diff
@@ -674,6 +674,49 @@ python3 $GS_HOME/tests/end2end-tests/check_infer.py --train-embout /data/gsgnn_m
...
+echo "**************[Multi-task] dataset: Movielens, RGCN layer 1, node feat: fixed HF BERT, BERT nodes: movie, inference with test"
+python3 -m graphstorm.run.gs_multi_task_learning --inference --workspace $GS_HOME/inference_scripts/mt_infer  --num-trainers $NUM_INFERs --num-servers 1 --num-samplers 0 --part-config /data/movielen_100k_multi_task_train_val_1p_4t/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_nc_lp_norm_with_mask_infer.yaml --use-mini-batch-infer false  --save-embed-path /data/gsgnn_mt/infer-emb/ --restore-model-path /data/gsgnn_mt/epoch-2 --save-prediction-path /data/gsgnn_mt/prediction/ --logging-file /tmp/infer_log.txt --preserve-input True
+
+error_and_exit $?
+
+cnt=$(ls -l /data/gsgnn_mt/infer-emb/ | wc -l)
+cnt=$[cnt - 2]
+if test $cnt != 4
+then
+     echo "The number of saved embs $cnt is not equal to 3. Two for movie and user and One for link-prediction-subtask normalized embedding."
```
Comment (zhjwy9343) [inline tests/end2end-tests/graphstorm-mt/mgpu_test.sh:709]:
```suggestion
     echo "The number of saved embs $cnt is not equal to 3. Should have Two for movie and user and One for link-prediction-subtask normalized embedding."
```

File: python/graphstorm/gconstruct/remap_result.py [hunk 5]
```diff
@@ -773,7 +787,8 @@ def main(args, gs_config_args):
 
         else: # There is no shared file system
             emb_names = os.listdir(node_emb_dir)
-            emb_names = [e_name for e_name in emb_names if e_name != "emb_info.json"]
+            emb_names = [e_name for e_name in emb_names \
+                if e_name not in task_emb_dirs + ["emb_info.json"]]
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:791]: The embeddings are saved as:
```
ntype0
ntype1
..
emb_info.json
task_xxx/
task_xxx/
```
task_xxx is in task_emb_dirs.

File: python/graphstorm/gconstruct/remap_result.py [hunk 6]
```diff
@@ -724,6 +737,7 @@ def main(args, gs_config_args):
         id_mapping_path = args.node_id_mapping
         predict_dir = args.prediction_dir
         node_emb_dir = args.node_emb_dir
+        task_emb_dirs = []
```
Comment (classicsong) [inline python/graphstorm/gconstruct/remap_result.py:740]: Add a commet

File: python/graphstorm/gconstruct/remap_result.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/gconstruct/remap_result.py:791]: added some coments.

**Review by zhjwy9343 — APPROVED** [review]
LGTM

**Pull request merged**